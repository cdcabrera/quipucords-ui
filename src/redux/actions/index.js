import { credentialsActions } from './credentialsActions';
import { factsActions } from './factsActions';
import { reportsActions } from './reportsActions';
import { scansActions } from './scansActions';
import { sourcesActions } from './sourcesActions';
import { statusActions } from './statusActions';
import { userActions } from './userActions';

const reduxActions = {
  credentials: credentialsActions,
  facts: factsActions,
  reports: reportsActions,
  scans: scansActions,
  sources: sourcesActions,
  status: statusActions,
  user: userActions
};

export {
  reduxActions as default,
  reduxActions,
  credentialsActions,
  factsActions,
  reportsActions,
  scansActions,
  sourcesActions,
  statusActions,
  userActions
};
