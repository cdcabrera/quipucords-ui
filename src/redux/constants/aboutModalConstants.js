const ABOUT_MODAL_SHOW = 'ABOUT_MODAL_SHOW';
const ABOUT_MODAL_HIDE = 'ABOUT_MODAL_HIDE';

const aboutModalTypes = {
  ABOUT_MODAL_SHOW,
  ABOUT_MODAL_HIDE
};

export { aboutModalTypes as default, aboutModalTypes, ABOUT_MODAL_SHOW, ABOUT_MODAL_HIDE };
