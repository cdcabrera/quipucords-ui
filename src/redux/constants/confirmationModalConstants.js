const CONFIRMATION_MODAL_SHOW = 'CONFIRMATION_MODAL_SHOW';
const CONFIRMATION_MODAL_HIDE = 'CONFIRMATION_MODAL_HIDE';

const confirmationModalTypes = {
  CONFIRMATION_MODAL_SHOW,
  CONFIRMATION_MODAL_HIDE
};

export { confirmationModalTypes as default, confirmationModalTypes, CONFIRMATION_MODAL_SHOW, CONFIRMATION_MODAL_HIDE };
