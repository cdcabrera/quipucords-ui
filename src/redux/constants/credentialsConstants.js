const ADD_CREDENTIAL = 'ADD_CREDENTIAL';
const DELETE_CREDENTIAL = 'DELETE_CREDENTIAL';
const GET_CREDENTIALS = 'GET_CREDENTIALS';
const UPDATE_CREDENTIAL = 'UPDATE_CREDENTIAL';
const CREATE_CREDENTIAL_SHOW = 'CREATE_CREDENTIAL_SHOW';
const EDIT_CREDENTIAL_SHOW = 'EDIT_CREDENTIAL_SHOW';
const UPDATE_CREDENTIAL_HIDE = 'UPDATE_CREDENTIAL_HIDE';

const SELECT_CREDENTIAL = 'SELECT_CREDENTIAL';
const DESELECT_CREDENTIAL = 'DESELECT_CREDENTIAL';
const EXPANDED_CREDENTIAL = 'EXPANDED_CREDENTIAL';
const NOT_EXPANDED_CREDENTIAL = 'NOT_EXPANDED_CREDENTIAL';
const UPDATE_CREDENTIALS = 'UPDATE_CREDENTIALS';

const credentialsTypes = {
  ADD_CREDENTIAL,
  UPDATE_CREDENTIAL,
  DELETE_CREDENTIAL,
  GET_CREDENTIALS,
  CREATE_CREDENTIAL_SHOW,
  EDIT_CREDENTIAL_SHOW,
  UPDATE_CREDENTIAL_HIDE,
  SELECT_CREDENTIAL,
  DESELECT_CREDENTIAL,
  EXPANDED_CREDENTIAL,
  NOT_EXPANDED_CREDENTIAL,
  UPDATE_CREDENTIALS
};

export {
  credentialsTypes as default,
  credentialsTypes,
  ADD_CREDENTIAL,
  UPDATE_CREDENTIAL,
  DELETE_CREDENTIAL,
  GET_CREDENTIALS,
  CREATE_CREDENTIAL_SHOW,
  EDIT_CREDENTIAL_SHOW,
  UPDATE_CREDENTIAL_HIDE,
  SELECT_CREDENTIAL,
  DESELECT_CREDENTIAL,
  EXPANDED_CREDENTIAL,
  NOT_EXPANDED_CREDENTIAL,
  UPDATE_CREDENTIALS
};
