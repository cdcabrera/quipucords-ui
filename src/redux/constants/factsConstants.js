const ADD_FACTS = 'ADD_FACTS';

const factsTypes = {
  ADD_FACTS
};

export { factsTypes as default, factsTypes, ADD_FACTS };
