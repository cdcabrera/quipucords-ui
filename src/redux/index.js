import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import store from './store';
import reduxActions from './actions';
import reduxReducers from './reducers';
import reduxSelectors from './selectors';
import reduxTypes from './constants';

const connectTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(withTranslation()(component));

const connectRouterTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connectTranslate(mapStateToProps, mapDispatchToProps)(component));

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(
    connect(
      mapStateToProps,
      mapDispatchToProps
    )(component)
  );

export {
  connect,
  connectRouter,
  connectRouterTranslate,
  connectTranslate,
  reduxActions,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store
};
