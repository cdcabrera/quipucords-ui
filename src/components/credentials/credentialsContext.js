import React, { useState } from 'react';
import { useShallowCompareEffect } from 'react-use';
import { AlertVariant, List, ListItem } from '@patternfly/react-core';
import { ContextIcon, ContextIconVariant } from '../contextIcon/contextIcon';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { apiTypes } from '../../constants/apiConstants';
import { helpers } from '../../common';
import { translate } from '../i18n/i18n';
import { useConfirmation } from '../../hooks/useConfirmation';

/**
 * Credential action, onDelete.
 *
 * @param {object} options
 * @param {Function} options.deleteCredentials
 * @param {Function} options.t
 * @param {Function} options.useConfirmation
 * @param {Function} options.useDispatch
 * @param {Function} options.useSelectorsResponse
 * @returns {(function(*): void)|*}
 */
const useOnDelete = ({
  deleteCredentials = reduxActions.credentials.deleteCredential,
  t = translate,
  useConfirmation: useAliasConfirmation = useConfirmation,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse
} = {}) => {
  const onConfirmation = useAliasConfirmation();
  const [credentialsToDelete, setCredentialsToDelete] = useState([]);
  const dispatch = useAliasDispatch();
  const {
    data,
    error: deletedError,
    fulfilled: deletedFulfilled
  } = useAliasSelectorsResponse(({ credentials }) => credentials?.deleted);
  const { errorMessage } = data?.[0] || {};

  useShallowCompareEffect(() => {
    if (credentialsToDelete.length) {
      const credentialIds = credentialsToDelete.map(cred => cred[apiTypes.API_RESPONSE_CREDENTIAL_ID]);
      deleteCredentials(credentialIds)(dispatch);
    }
  }, [credentialsToDelete, deleteCredentials, dispatch]);

  useShallowCompareEffect(() => {
    if (deletedFulfilled && credentialsToDelete.length) {
      const credentialNames = credentialsToDelete.map(cred => cred[apiTypes.API_RESPONSE_CREDENTIAL_NAME]);
      dispatch([
        {
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: AlertVariant.success,
          header: t('toast-notifications.title', {
            context: ['deleted-credential'],
            count: credentialNames.length
          }),
          message: t('toast-notifications.description', {
            context: ['deleted-credential'],
            name: credentialNames[0],
            count: credentialNames.length
          })
        },
        {
          type: reduxTypes.credentials.DESELECT_CREDENTIAL,
          item: credentialsToDelete
        },
        {
          type: reduxTypes.credentials.UPDATE_CREDENTIALS
        }
      ]);

      setCredentialsToDelete(() => []);
    }

    if (deletedError && credentialsToDelete.length) {
      const credentialNames = credentialsToDelete.map(cred => cred[apiTypes.API_RESPONSE_CREDENTIAL_NAME]);
      dispatch([
        {
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: AlertVariant.danger,
          header: t('toast-notifications.title', {
            context: ['error']
          }),
          message: t('toast-notifications.description', {
            context: ['deleted-credential', 'error'],
            name: credentialNames[0],
            count: credentialNames.length,
            message: errorMessage
          })
        },
        {
          type: reduxTypes.credentials.DESELECT_CREDENTIAL,
          item: credentialsToDelete
        },
        {
          type: reduxTypes.credentials.UPDATE_CREDENTIALS
        }
      ]);

      setCredentialsToDelete(() => []);
    }
  }, [credentialsToDelete, deletedError, deletedFulfilled, dispatch, errorMessage, t]);

  /**
   * Confirmation, and state, for deleting a single or multiple credentials
   *
   * @param {Array|string} credentials
   */
  return credentials => {
    const updatedCredentials = (Array.isArray(credentials) && credentials) || [credentials];

    onConfirmation({
      title: t('form-dialog.confirmation', {
        context: ['title', 'delete-credential'],
        count: updatedCredentials.length
      }),
      heading: t(
        'form-dialog.confirmation',
        {
          context: ['heading', 'delete-credential'],
          count: updatedCredentials.length,
          name: updatedCredentials?.[0]?.[apiTypes.API_RESPONSE_CREDENTIAL_NAME]
        },
        [<strong />]
      ),
      body:
        (updatedCredentials.length > 1 && (
          <List className="quipucords-list__overflow-scroll" isPlain>
            {updatedCredentials.map(
              ({
                [apiTypes.API_RESPONSE_CREDENTIAL_NAME]: name,
                [apiTypes.API_RESPONSE_CREDENTIAL_CRED_TYPE]: credType
              }) => (
                <ListItem key={name} icon={<ContextIcon symbol={ContextIconVariant[credType]} />}>
                  {name}
                </ListItem>
              )
            )}
          </List>
        )) ||
        undefined,
      confirmButtonText: t('form-dialog.label', {
        context: ['delete']
      }),
      onConfirm: () => setCredentialsToDelete(() => updatedCredentials)
    });
  };
};

/**
 * On edit a credential, show modal.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnEdit = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return item => {
    dispatch({
      type: reduxTypes.credentials.EDIT_CREDENTIAL_SHOW,
      credential: item
    });
  };
};

/**
 * On expand a row facet.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnExpand = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ isExpanded, cellIndex, data }) => {
    dispatch({
      type: isExpanded ? reduxTypes.credentials.EXPANDED_CREDENTIAL : reduxTypes.credentials.NOT_EXPANDED_CREDENTIAL,
      viewType: reduxTypes.view.CREDENTIALS_VIEW,
      item: data.item,
      cellIndex
    });
  };
};

/**
 * On refresh view.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnRefresh = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return () => {
    dispatch({
      type: reduxTypes.credentials.UPDATE_CREDENTIALS
    });
  };
};

/**
 * On select a row.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnSelect = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ isSelected, data }) => {
    dispatch({
      type: isSelected ? reduxTypes.credentials.SELECT_CREDENTIAL : reduxTypes.credentials.DESELECT_CREDENTIAL,
      viewType: reduxTypes.view.CREDENTIALS_VIEW,
      item: data.item
    });
  };
};

/**
 * Get credentials
 *
 * @param {object} options
 * @param {Function} options.getCredentials
 * @param {Function} options.useDispatch
 * @param {Function} options.useSelectors
 * @param {Function} options.useSelectorsResponse
 * @returns {{date: *, data: *[], pending: boolean, errorMessage: null, fulfilled: boolean, selectedRows: *,
 *     expandedRows: *, error: boolean}}
 */
const useGetCredentials = ({
  getCredentials = reduxActions.credentials.getCredentials,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useSelectors: useAliasSelectors = storeHooks.reactRedux.useSelectors,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse
} = {}) => {
  const dispatch = useAliasDispatch();
  const [refreshUpdate, selectedRows, expandedRows, viewOptions] = useAliasSelectors([
    ({ credentials }) => credentials?.update,
    ({ credentials }) => credentials?.selected,
    ({ credentials }) => credentials?.expanded,
    ({ viewOptions: stateViewOptions }) => stateViewOptions?.[reduxTypes.view.CREDENTIALS_VIEW]
  ]);
  const {
    data: responseData,
    error,
    fulfilled,
    message: errorMessage,
    pending,
    responses = {}
  } = useAliasSelectorsResponse({ id: 'view', selector: ({ credentials }) => credentials?.view });

  const [{ date } = {}] = responses?.list || [];
  const { results: data = [] } = responseData?.view || {};
  const query = helpers.createViewQueryObject(viewOptions);

  useShallowCompareEffect(() => {
    getCredentials(null, query)(dispatch);
  }, [dispatch, getCredentials, query, refreshUpdate]);

  return {
    pending,
    error,
    errorMessage,
    fulfilled,
    data,
    date,
    selectedRows,
    expandedRows
  };
};

const context = {
  useGetCredentials,
  useOnDelete,
  useOnEdit,
  useOnExpand,
  useOnRefresh,
  useOnSelect
};

export {
  context as default,
  context,
  useGetCredentials,
  useOnDelete,
  useOnEdit,
  useOnExpand,
  useOnRefresh,
  useOnSelect
};
