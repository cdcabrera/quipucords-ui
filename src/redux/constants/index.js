import { aboutModalTypes } from './aboutModalConstants';
import { confirmationModalTypes } from './confirmationModalConstants';
import { credentialsTypes } from './credentialsConstants';
import { factsTypes } from './factsConstants';
import { reportsTypes } from './reportsConstants';
import { scansTypes } from './scansConstants';
import { sourcesTypes } from './sourcesConstants';
import { statusTypes } from './statusConstants';
import { toastNotificationTypes } from './toasNotificationConstants';
import { userTypes } from './userConstants';
import { viewTypes } from './viewConstants';
import { viewPaginationTypes } from './viewPaginationConstants';
import { viewToolbarTypes } from './viewToolbarConstants';

const reduxTypes = {
  aboutModal: aboutModalTypes,
  confirmationModal: confirmationModalTypes,
  credentials: credentialsTypes,
  facts: factsTypes,
  reports: reportsTypes,
  scans: scansTypes,
  sources: sourcesTypes,
  status: statusTypes,
  toastNotifications: toastNotificationTypes,
  user: userTypes,
  view: viewTypes,
  viewPagination: viewPaginationTypes,
  viewToolbar: viewToolbarTypes
};

export {
  reduxTypes as default,
  reduxTypes,
  aboutModalTypes,
  confirmationModalTypes,
  credentialsTypes,
  factsTypes,
  reportsTypes,
  scansTypes,
  sourcesTypes,
  statusTypes,
  toastNotificationTypes,
  userTypes,
  viewTypes,
  viewPaginationTypes,
  viewToolbarTypes
};
