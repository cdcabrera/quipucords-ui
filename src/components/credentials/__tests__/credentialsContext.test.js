import { context, useGetCredentials, useOnDelete, useOnEdit } from '../credentialsContext';
import { apiTypes } from '../../../constants/apiConstants';
import { reduxTypes } from '../../../redux';

describe('CredentialsContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should handle deleting a credential with a confirmation', async () => {
    const mockConfirmation = jest.fn();
    const mockCredential = {
      [apiTypes.API_RESPONSE_CREDENTIAL_ID]: 'lorem ipsum base id',
      [apiTypes.API_RESPONSE_CREDENTIAL_NAME]: 'lorem ipsum name'
    };

    const { result } = await shallowHook(() =>
      useOnDelete({
        useConfirmation: () => mockConfirmation
      })
    );

    result(mockCredential);

    expect(mockConfirmation.mock.calls).toMatchSnapshot('dispatch onDelete');
    mockConfirmation.mockClear();
  });

  it('should handle credential actions for onEdit', async () => {
    const mockDispatch = jest.fn();
    const mockCredential = {
      [apiTypes.API_RESPONSE_CREDENTIAL_ID]: 'lorem ipsum base id',
      [apiTypes.API_RESPONSE_CREDENTIAL_NAME]: 'lorem ipsum name'
    };

    const { result } = await shallowHook(() =>
      useOnEdit({
        useDispatch: () => mockDispatch
      })
    );

    result(mockCredential);

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch onEdit');
    mockDispatch.mockClear();
  });

  it('should apply a hook for retrieving data from multiple selectors', () => {
    const { result: errorResponse } = shallowHook(() =>
      useGetCredentials({
        useSelectorsResponse: () => ({ error: true, message: 'Lorem ipsum' })
      })
    );

    const { result: pendingResponse } = shallowHook(() =>
      useGetCredentials({
        useSelectorsResponse: () => ({ pending: true })
      })
    );

    const { result: fulfilledResponse } = shallowHook(() =>
      useGetCredentials({
        useSelectorsResponse: () => ({ fulfilled: true, data: { view: { results: ['dolor', 'sit'] } } })
      })
    );

    const { result: mockStoreSuccessResponse } = shallowHook(() => useGetCredentials(), {
      state: {
        viewOptions: {
          [reduxTypes.view.SCANS_VIEW]: {}
        },
        credentials: {
          expanded: {},
          selected: {},
          view: {
            fulfilled: true,
            data: {
              results: ['lorem', 'ipsum']
            }
          }
        }
      }
    });

    expect({ errorResponse, fulfilledResponse, pendingResponse, mockStoreSuccessResponse }).toMatchSnapshot(
      'responses'
    );
  });
});
