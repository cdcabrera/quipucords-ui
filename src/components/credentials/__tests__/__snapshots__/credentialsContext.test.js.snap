// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialsContext should apply a hook for retrieving data from multiple selectors: responses 1`] = `
Object {
  "errorResponse": Object {
    "data": Array [],
    "date": undefined,
    "error": true,
    "errorMessage": "Lorem ipsum",
    "expandedRows": undefined,
    "fulfilled": undefined,
    "hasData": false,
    "pending": undefined,
    "selectedRows": undefined,
    "totalResults": 0,
  },
  "fulfilledResponse": Object {
    "data": Array [
      "dolor",
      "sit",
    ],
    "date": undefined,
    "error": undefined,
    "errorMessage": undefined,
    "expandedRows": undefined,
    "fulfilled": true,
    "hasData": false,
    "pending": undefined,
    "selectedRows": undefined,
    "totalResults": 0,
  },
  "mockStoreSuccessResponse": Object {
    "data": Array [
      "lorem",
      "ipsum",
    ],
    "date": undefined,
    "error": false,
    "errorMessage": null,
    "expandedRows": Object {},
    "fulfilled": true,
    "hasData": false,
    "pending": false,
    "selectedRows": Object {},
    "totalResults": 0,
  },
  "pendingResponse": Object {
    "data": Array [],
    "date": undefined,
    "error": undefined,
    "errorMessage": undefined,
    "expandedRows": undefined,
    "fulfilled": undefined,
    "hasData": false,
    "pending": true,
    "selectedRows": undefined,
    "totalResults": 0,
  },
}
`;

exports[`CredentialsContext should handle credential actions for onEdit: dispatch onEdit 1`] = `
Array [
  Array [
    Object {
      "credential": Object {
        "id": "lorem ipsum base id",
        "name": "lorem ipsum name",
      },
      "type": "EDIT_CREDENTIAL_SHOW",
    },
  ],
]
`;

exports[`CredentialsContext should handle deleting a credential with a confirmation: dispatch onDelete 1`] = `
Array [
  Array [
    Object {
      "body": undefined,
      "confirmButtonText": "t(form-dialog.label, {\\"context\\":\\"delete\\"})",
      "heading": "t(form-dialog.confirmation_heading, {\\"context\\":\\"delete-credential\\",\\"count\\":1,\\"name\\":\\"lorem ipsum name\\"}, [object Object])",
      "onConfirm": [Function],
      "title": "t(form-dialog.confirmation_title, {\\"context\\":\\"delete-credential\\",\\"count\\":1})",
    },
  ],
]
`;

exports[`CredentialsContext should return specific properties: specific properties 1`] = `
Object {
  "INITIAL_QUERY": Object {
    "ordering": "name",
    "page": 1,
    "page_size": 10,
  },
  "VIEW_ID": "credentials",
  "useContextGetCredentials": [Function],
  "useGetCredentials": [Function],
  "useOnDelete": [Function],
  "useOnEdit": [Function],
  "useOnExpand": [Function],
  "useOnRefresh": [Function],
  "useOnSelect": [Function],
}
`;
